torch/optim/adagrad.py

```diff
@@ -66,13 +79,16 @@ def step(self, closure=None):
 
                 state['step'] += 1
 
-                if group['weight_decay'] != 0:
+                if group['weight_decay'] != 0 and group['l2_reg']:
                     if p.grad.data.is_sparse:
-                        raise RuntimeError("weight_decay option is not compatible with sparse gradients")
-                    grad = grad.add(group['weight_decay'], p.data)
+                        raise RuntimeError("L2 weight regularization option is not compatible with sparse gradients")
+                    grad.add_(group['weight_decay'], p.data)
 
                 clr = group['lr'] / (1 + (state['step'] - 1) * group['lr_decay'])
 
+                if group['weight_decay'] != 0 and not group['l2_reg']:
+                    p.data.add_(-group['weight_decay'], p.data)
```

Comment:
```suggestion
                    p.data.add_(-group['weight_decay'] * group['lr'], p.data)
```

In L2, the weight decay is stored inside the grad variable, which is later multiplied by the learning rate. This new formulation is not multiplying it by the learning rate (see Algorithms 1 and 2 in the paper).